commit 0b1007c3578569469a6fab6ae5cca918ccdc3ee1
Author: Tejun Heo <tj@kernel.org>
Date:   Thu Jun 2 11:13:59 2011 +0200

    ptrace: remove silly wait_trap variable from ptrace_attach()
    
    Remove local variable wait_trap which determines whether to wait for
    !TRAPPING or not and simply wait for it if attach was successful.
    
    -v2: Oleg pointed out wait should happen iff attach was successful.
    
    Signed-off-by: Tejun Heo <tj@kernel.org>
    Cc: Oleg Nesterov <oleg@redhat.com>
    Signed-off-by: Oleg Nesterov <oleg@redhat.com>
    [dannf: backported to Debian's 2.6.32]

diff --git a/kernel/ptrace.c b/kernel/ptrace.c
index c17536a..319dd93 100644
--- a/kernel/ptrace.c
+++ b/kernel/ptrace.c
@@ -183,7 +183,6 @@ bool ptrace_may_access(struct task_struct *task, unsigned int mode)
 
 int ptrace_attach(struct task_struct *task)
 {
-	bool wait_trap = false;
 	int retval;
 
 	audit_ptrace(task);
@@ -245,7 +244,6 @@ int ptrace_attach(struct task_struct *task)
 	if (task_is_stopped(task)) {
 		task->group_stop |= GROUP_STOP_PENDING | GROUP_STOP_TRAPPING;
 		signal_wake_up(task, 1);
-		wait_trap = true;
 	}
 
 	spin_unlock(&task->sighand->siglock);
@@ -256,7 +254,7 @@ unlock_tasklist:
 unlock_creds:
 	mutex_unlock(&task->cred_guard_mutex);
 out:
-	if (wait_trap)
+	if (!retval)
 		wait_event(current->signal->wait_chldexit,
 			   !(task->group_stop & GROUP_STOP_TRAPPING));
 	return retval;
